propolis/src/block/crucible.rs

```diff
@@ -0,0 +1,341 @@
+//! Implement a virtual block device backed by Crucible
+
+use std::collections::VecDeque;
+use std::io::{Error, ErrorKind, Result};
+use std::net::SocketAddrV4;
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Condvar, Mutex};
+
+use super::DeviceInfo;
+use crate::block;
+use crate::dispatch::{AsyncCtx, DispCtx, Dispatcher, SyncCtx, WakeFn};
+use crate::inventory::Entity;
+use crate::vmm::SubMapping;
+
+use crucible::CrucibleError;
+
+use tokio::sync::Semaphore;
+
+/// Helper function, because Rust couldn't derive the types
+fn map_crucible_error_to_io(x: CrucibleError) -> std::io::Error {
+    x.into()
+}
+
+pub struct CrucibleBackend {
+    guest: Arc<crucible::Guest>,
+    block_size: u64,
+    total_size: u64,
+    sectors: u64,
+    read_only: bool,
+
+    driver: Mutex<Option<Arc<SyncDriver>>>,
+}
+
+impl CrucibleBackend {
+    pub fn create(
+        disp: &Dispatcher,
+        targets: Vec<SocketAddrV4>,
+        read_only: bool,
+        key: Option<String>,
+    ) -> Result<Arc<Self>> {
+        CrucibleBackend::_create(disp, targets, read_only, key)
+            .map_err(map_crucible_error_to_io)
+    }
+
+    fn _create(
+        disp: &Dispatcher,
+        targets: Vec<SocketAddrV4>,
+        read_only: bool,
+        key: Option<String>,
+    ) -> anyhow::Result<Arc<Self>, crucible::CrucibleError> {
```

Comment:
Per the [rust crate zoo](https://github.com/oxidecomputer/meta/blob/master/engineering/rust-crate-zoo.adoc#anyhow) I'd been lead to believe that `thiserror` was generally to be used in libraries while `anyhow` was for applications.  Is crucible planning to follow that pattern? (Here, propolis is fulfilling the library role, so I'd expect us to use `thiserror` once we start working on the whole error situation)

Reply:
Sounds good.  In the mean time, I don't think you need to hold this PR up for oxidecomputer/crucible#85.